archunit/src/main/java/com/tngtech/archunit/library/freeze/ViolationStoreFactory.java

```diff
@@ -0,0 +1,225 @@
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+class ViolationStoreFactory {
+    static final String FREEZE_STORE_PROPERTY = "freeze.store";
+
+    static ViolationStore create() {
+        return ArchConfiguration.get().containsProperty(FREEZE_STORE_PROPERTY)
+                ? createInstance(ArchConfiguration.get().getProperty(FREEZE_STORE_PROPERTY))
+                : new TextFileBasedViolationStore();
+    }
+
+    @MayResolveTypesViaReflection(reason = "This is not part of the import process")
+    private static ViolationStore createInstance(String violationStoreClassName) {
+        try {
+            return (ViolationStore) newInstanceOf(Class.forName(violationStoreClassName));
+        } catch (Exception e) {
+            String message = String.format("Could not instantiate %s of configured type '%s=%s'",
+                    ViolationStore.class.getSimpleName(), FREEZE_STORE_PROPERTY, violationStoreClassName);
+            throw new StoreInitializationFailedException(message, e);
+        }
+    }
+
+    @VisibleForTesting
+    static class TextFileBasedViolationStore implements ViolationStore {
+        private static final Logger log = LoggerFactory.getLogger(TextFileBasedViolationStore.class);
+
+        private static final Pattern UNESCAPED_LINE_BREAK_PATTERN = Pattern.compile("(?<!\\\\)\n");
+        private File storeFolder;
+        private FileSyncedProperties storedRules;
+
+        @Override
+        public void initialize(Properties properties) {
+            String path = properties.getProperty("default.path", "archunit_store");
+            storeFolder = new File(path);
+            ensureExistence(storeFolder);
+            File storedRulesFile = new File(storeFolder, "stored.rules");
+            log.info("Initializing {} at {}", TextFileBasedViolationStore.class.getSimpleName(), storedRulesFile.getAbsolutePath());
+            storedRules = new FileSyncedProperties(storedRulesFile);
+            checkInitialization(storedRules.initializationSuccessful(), "Cannot create rule store at %s", storedRulesFile.getAbsolutePath());
+        }
+
+        private void ensureExistence(File folder) {
+            checkState(folder.exists() && folder.isDirectory() || folder.mkdirs(), "Cannot create folder %s", folder.getAbsolutePath());
+        }
+
+        private void checkInitialization(boolean initializationSuccessful, String message, Object... args) {
+            if (!initializationSuccessful) {
+                throw new StoreInitializationFailedException(String.format(message, args));
+            }
+        }
+
+        @Override
+        public boolean contains(ArchRule rule) {
+            return storedRules.containsKey(rule.getDescription());
+        }
+
+        @Override
+        public void save(ArchRule rule, List<String> violations) {
+            log.debug("Storing evaluated rule '{}' with violations: {}", rule.getDescription(), violations);
```

Comment:
Include number of violations?
Same for `getViolations` below.

Reply:
Thumbs up for all your include the number comments :wink: